### PR TITLE
flex teaser block

### PR DESCRIPTION
**Pull request by penk85**
Editor: [here](https://editor.fwpc51fle.wepublish.dev/)
Admin API: [here](https://api.fwpc51fle.wepublish.dev/admin)
Public API: [here](https://api.fwpc51fle.wepublish.dev/)

### CODE REVIEW

File: packages/editor/package.json [hunk 1]
```diff
@@ -52,6 +52,7 @@
     "@apollo/client": "^3.3.21",
     "@karma.run/react": "^0.1.3",
     "@karma.run/webpack": "^0.0.3",
+    "@types/react-grid-layout": "^1.1.2",
```
Comment (bishoymelek) [inline packages/editor/package.json:55]: Can you move this to devDeps?

File: packages/editor/src/client/blocks/teaserFlexGridBlock.tsx [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+import {BlockProps} from '../atoms/blockList'
+import nanoid from 'nanoid'
+import {ButtonToolbar, Drawer, Icon, IconButton, Panel} from 'rsuite'
+import {IconButtonTooltip} from '../atoms/iconButtonTooltip'
+import {contentForTeaser, TeaserBlockProps} from './teaserGridBlock'
+import {PlaceholderInput} from '../atoms/placeholderInput'
+import {TeaserEditPanel} from '../panel/teaserEditPanel'
+import {TeaserSelectAndEditPanel} from '../panel/teaserSelectAndEditPanel'
+
+// import {useTranslation} from 'react-i18next'
```
Comment (bishoymelek) [inline packages/editor/src/client/blocks/teaserFlexGridBlock.tsx:17]: can you clean this one?

File: packages/editor/src/client/blocks/teaserFlexGridBlock.tsx [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+            <div
+              style={{
+                position: 'absolute',
+                zIndex: 1,
+                right: 0,
+                top: 0
+              }}>
+              {/* eslint-disable-next-line i18next/no-literal-string */}
+              <IconButtonTooltip caption="choose teaser">
```
Comment (bishoymelek) [inline packages/editor/src/client/blocks/teaserFlexGridBlock.tsx:53]: Don't we have these strings in translation files already? 👀 if not I would recommend adding them, what do you think?

File: packages/editor/src/client/blocks/teaserFlexGridBlock.tsx [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    setFlexTeasers(
+      flexTeasers.map(flexTeaser => {
+        return flexTeaser.alignment.i === index
+          ? {alignment: flexTeaser.alignment, teaser: null}
+          : flexTeaser
+      })
+    )
+  }
+
+  // @ts-ignore
```
Comment (bishoymelek) [inline packages/editor/src/client/blocks/teaserFlexGridBlock.tsx:186]: do you need ignore statements here?

File: packages/editor/src/client/blocks/teaserGridBlock.tsx [hunk 5]
```diff
@@ -131,7 +131,7 @@ export function TeaserGridBlock({value, onChange}: BlockProps<TeaserGridBlockVal
 export interface TeaserBlockProps {
   teaser: Teaser | null
   showGrabCursor: boolean
-  numColumns: number
+  numColumns?: number
```
Comment (bishoymelek) [inline packages/editor/src/client/blocks/teaserGridBlock.tsx:134]: why did you change in that file? I think not related?

**Review by bishoymelek — CHANGES_REQUESTED** [review]
Great work @penk85 🎉 very small cleanup points

File: packages/editor/src/client/blocks/teaserFlexGridBlock.tsx (hunk 2 above)
Comment (penk85) [reply packages/editor/src/client/blocks/teaserFlexGridBlock.tsx:17]: @bishoymelek the translation import? I'll use it.

File: packages/editor/src/client/blocks/teaserFlexGridBlock.tsx (hunk 3 above)
Comment (penk85) [reply packages/editor/src/client/blocks/teaserFlexGridBlock.tsx:53]: Yes. That'll be moved to translation later.

File: packages/editor/src/client/blocks/teaserGridBlock.tsx (hunk 5 above)
Comment (penk85) [reply packages/editor/src/client/blocks/teaserGridBlock.tsx:134]: @bishoymelek I wanted to be able to use the existing teaser functions. Essentially the same thing besides the columns.

**Review by camilleory — DISMISSED** [review]

File: examples/website/src/shared/blocks/blockRenderer.tsx [hunk 6]
```diff
@@ -142,7 +146,7 @@ export function renderBlock(block: Block | null, opts: RenderBlockOptions) {
   }
 }
 
-function renderTeaser(key: string, article: PublishedArticle, isPeerArticle = false) {
+export function renderTeaser(key: string, article: PublishedArticle, isPeerArticle = false) {
```
Comment (bishoymelek) [inline examples/website/src/shared/blocks/blockRenderer.tsx:149]: Do you need this? 👀

File: examples/website/src/shared/types.ts [hunk 7]
```diff
@@ -334,6 +335,27 @@ export type GridBlock = BaseBlock<
...
+  x: number
+  y: number
+  w: number
+  h: number
+  static?: boolean
+}
+
+export interface FlexTeaser {
+  alignment: FlexItemAlignment
+  blocks: Block[]
```
Comment (bishoymelek) [inline examples/website/src/shared/types.ts:349]: How can a Teaser will have multiple blocks?  And is the key `blocks` or `teaser` ?

File: packages/editor/src/client/blocks/teaserGridBlock.tsx (hunk 5 above)
Comment (bishoymelek) [reply packages/editor/src/client/blocks/teaserGridBlock.tsx:134]: @penk85 but isn't that affecting other places where this interface already being used?

**Review by bishoymelek — CHANGES_REQUESTED** [review]
Great work @penk85 🎉🎉 I have few questions.

File: examples/website/src/shared/blocks/blockRenderer.tsx (hunk 6 above)
Comment (penk85) [reply examples/website/src/shared/blocks/blockRenderer.tsx:149]: Actually not anymore

File: examples/website/src/shared/types.ts (hunk 7 above)
Comment (penk85) [reply examples/website/src/shared/types.ts:349]: There's a teaser in the blocks. It was a workaround to use the existing teaser render functions. Working on finding a better solution.

**Review by tomaszdurka — APPROVED** [review]
Good job Penina.

It works pretty well in admin and in api - these I've tested.

One questions - I understand static is not really needed for public api, but only for private/admin usage, right?

```
# Write your query or mutation here
query {
  articles(first: 10){
    nodes{
      id
      blocks {
        __typename
        ... on TeaserFlexGridBlock {
          flexTeasers {
            alignment {
              i
              x
              y
              w
              h
              static
            }
            teaser{
              __typename
            }
          }
        }
      }
    }
  }
  
}
```

**Review by camilleory — APPROVED** [review]

File: packages/api/src/db/block.ts [hunk 8]
```diff
@@ -191,6 +192,25 @@ export interface TeaserGridBlock {
   numColumns: number
 }
 
+export interface FlexItemAlignment {
```
Comment (roosnic1) [inline packages/api/src/db/block.ts:195]: I'm a bit confused by the name. Why does it have `Item` in it. I think `FlexAlignment` would be better here.

File: packages/api/src/db/block.ts [hunk 9]
```diff
@@ -191,6 +192,25 @@ export interface TeaserGridBlock {
...
+  h: number
+  static?: boolean
+}
+
+export interface FlexTeaser {
+  alignment: FlexItemAlignment
+  teaser: Teaser | null
+}
+
+export interface TeaserFlexGridBlock {
```
Comment (roosnic1) [inline packages/api/src/db/block.ts:209]: also the naming. Try to stay consistent. `TeaserGridFlexBlock`

File: packages/api/src/graphql/blocks.ts [hunk 10]
```diff
@@ -176,6 +178,48 @@ export const GraphQLTeaserGridBlock = new GraphQLObjectType<TeaserGridBlock, Con
...
 
+export const GraphQLFlexGridItemAlignment = new GraphQLObjectType<FlexItemAlignment, Context>({
+  name: 'FlexItemAlignment',
+  fields: {
+    i: {type: GraphQLNonNull(GraphQLString)},
+    x: {type: GraphQLNonNull(GraphQLInt)},
+    y: {type: GraphQLNonNull(GraphQLInt)},
+    w: {type: GraphQLNonNull(GraphQLInt)},
+    h: {type: GraphQLNonNull(GraphQLInt)},
+    static: {type: GraphQLBoolean}
```
Comment (roosnic1) [inline packages/api/src/graphql/blocks.ts:189]: this one should also not be optional

File: packages/api/src/graphql/blocks.ts [hunk 11]
```diff
@@ -176,6 +178,48 @@ export const GraphQLTeaserGridBlock = new GraphQLObjectType<TeaserGridBlock, Con
   })
 })
 
+export const GraphQLFlexGridItemAlignment = new GraphQLObjectType<FlexItemAlignment, Context>({
+  name: 'FlexItemAlignment',
```
Comment (roosnic1) [inline packages/api/src/graphql/blocks.ts:182]: Also remove the `Item` from the name

File: packages/api/src/graphql/blocks.ts [hunk 12]
```diff
@@ -176,6 +178,48 @@ export const GraphQLTeaserGridBlock = new GraphQLObjectType<TeaserGridBlock, Con
...
+    x: {type: GraphQLNonNull(GraphQLInt)},
+    y: {type: GraphQLNonNull(GraphQLInt)},
+    w: {type: GraphQLNonNull(GraphQLInt)},
+    h: {type: GraphQLNonNull(GraphQLInt)},
+    static: {type: GraphQLBoolean}
+  }
+})
+
+export const GraphQLFlexGridItemAlignmentInput = new GraphQLInputObjectType({
+  name: 'FlexItemAlignmentInput',
```
Comment (roosnic1) [inline packages/api/src/graphql/blocks.ts:194]: remove the `Item` from the name

File: packages/api/src/graphql/blocks.ts [hunk 13]
```diff
@@ -176,6 +178,48 @@ export const GraphQLTeaserGridBlock = new GraphQLObjectType<TeaserGridBlock, Con
...
+  fields: {
+    i: {type: GraphQLNonNull(GraphQLString)},
+    x: {type: GraphQLNonNull(GraphQLInt)},
+    y: {type: GraphQLNonNull(GraphQLInt)},
+    w: {type: GraphQLNonNull(GraphQLInt)},
+    h: {type: GraphQLNonNull(GraphQLInt)},
+    static: {type: GraphQLBoolean}
```
Comment (roosnic1) [inline packages/api/src/graphql/blocks.ts:201]: not optional.

File: packages/api/src/graphql/blocks.ts [hunk 14]
```diff
@@ -176,6 +178,48 @@ export const GraphQLTeaserGridBlock = new GraphQLObjectType<TeaserGridBlock, Con
...
+export const GraphQLFlexTeaser = new GraphQLObjectType<FlexTeaser, Context>({
+  name: 'FlexTeaser',
+  fields: {
+    alignment: {type: GraphQLNonNull(GraphQLFlexGridItemAlignment)},
+    teaser: {type: GraphQLTeaser}
+  }
+})
+
+export const GraphQLTeaserFlexGridBlock = new GraphQLObjectType<TeaserFlexGridBlock, Context>({
+  name: 'TeaserFlexGridBlock',
```
Comment (roosnic1) [inline packages/api/src/graphql/blocks.ts:214]: Also the naming `GraphQLTeaserGridFlexBlock` and  `TeaserGridFlexBlock`

File: packages/api/src/graphql/blocks.ts [hunk 15]
```diff
@@ -288,6 +332,40 @@ export const GraphQLPublicTeaserGridBlock = new GraphQLObjectType<TeaserGridBloc
   })
 })
 
+export const GraphQLFlexGridItemPublicAlignment = new GraphQLObjectType<FlexItemAlignment, Context>(
+  {
+    name: 'FlexItemAlignment',
```
Comment (roosnic1) [inline packages/api/src/graphql/blocks.ts:337]: remote the `item`

File: packages/api/src/graphql/blocks.ts [hunk 16]
```diff
@@ -288,6 +332,40 @@ export const GraphQLPublicTeaserGridBlock = new GraphQLObjectType<TeaserGridBloc
...
+  name: 'FlexTeaser',
+  fields: {
+    alignment: {type: GraphQLNonNull(GraphQLFlexGridItemPublicAlignment)},
+    teaser: {type: GraphQLPublicTeaser}
+  }
+})
+
+export const GraphQLPublicTeaserFlexGridBlock = new GraphQLObjectType<TeaserFlexGridBlock, Context>(
+  {
+    name: 'TeaserFlexGridBlock',
```
Comment (roosnic1) [inline packages/api/src/graphql/blocks.ts:357]: naming

File: packages/api/src/graphql/blocks.ts [hunk 17]
```diff
@@ -720,6 +798,21 @@ export const GraphQLTeaserGridBlockInput = new GraphQLInputObjectType({
...
+export const GraphQLFlexTeaserInput = new GraphQLInputObjectType({
+  name: 'FlexTeaserInput',
+  fields: {
+    teaser: {type: GraphQLTeaserInput},
+    alignment: {type: GraphQLNonNull(GraphQLFlexGridItemAlignmentInput)}
+  }
+})
+
+export const GraphQLTeaserFlexGridBlockInput = new GraphQLInputObjectType({
+  name: 'TeaserFlexGridBlockInput',
```
Comment (roosnic1) [inline packages/api/src/graphql/blocks.ts:810]: naming

File: packages/editor/src/client/blocks/types.ts [hunk 18]
```diff
@@ -209,6 +210,24 @@ export interface TeaserGridBlockValue {
   numColumns: number
 }
 
+export interface FlexItemAlignment {
```
Comment (roosnic1) [inline packages/editor/src/client/blocks/types.ts:213]: naming

File: packages/editor/src/client/blocks/types.ts [hunk 19]
```diff
@@ -209,6 +210,24 @@ export interface TeaserGridBlockValue {
...
+  i: string
+  x: number
+  y: number
+  w: number
+  h: number
+  static?: boolean
+}
+
+export interface FlexTeaser {
+  alignment: FlexItemAlignment
```
Comment (roosnic1) [inline packages/editor/src/client/blocks/types.ts:223]: naming

File: packages/editor/src/client/blocks/types.ts [hunk 20]
```diff
@@ -209,6 +210,24 @@ export interface TeaserGridBlockValue {
...
+  teaser: Teaser | null
+}
+
+export interface TeaserFlexGridBlockValue {
```
Comment (roosnic1) [inline packages/editor/src/client/blocks/types.ts:227]: naming

File: packages/editor/src/client/blocks/types.ts [hunk 21]
```diff
@@ -392,6 +417,91 @@ export function unionMapForBlock(block: BlockValue): BlockInput {
...
+                      articleID: flexTeaser.teaser.article.id
+                    }
+                  },
+                  alignment: {
+                    i: flexTeaser?.alignment.i ?? nanoid(),
+                    x: flexTeaser?.alignment.x ?? 1,
+                    y: flexTeaser?.alignment.y ?? 1,
+                    w: flexTeaser?.alignment.w ?? 1,
+                    h: flexTeaser?.alignment.h ?? 1,
+                    static: flexTeaser?.alignment.static ?? false
```
Comment (roosnic1) [inline packages/editor/src/client/blocks/types.ts:443]: i think you don't need the `?` and the `??` here. TypeScript should know about those values.

File: packages/editor/src/client/blocks/types.ts [hunk 22]
```diff
@@ -392,6 +417,91 @@ export function unionMapForBlock(block: BlockValue): BlockInput {
...
+                    peerID: flexTeaser.teaser.peer.id,
+                    articleID: flexTeaser.teaser.articleID
+                  },
+                  alignment: {
+                    i: flexTeaser?.alignment.i ?? nanoid(),
+                    x: flexTeaser?.alignment.x ?? 1,
+                    y: flexTeaser?.alignment.y ?? 1,
+                    w: flexTeaser?.alignment.w ?? 1,
+                    h: flexTeaser?.alignment.h ?? 1,
+                    static: flexTeaser?.alignment.static ?? false
```
Comment (roosnic1) [inline packages/editor/src/client/blocks/types.ts:464]: i think you don't need the `?` and the `??` here. TypeScript should know about those values.

File: packages/editor/src/client/blocks/types.ts [hunk 23]
```diff
@@ -392,6 +417,91 @@ export function unionMapForBlock(block: BlockValue): BlockInput {
...
+                    lead: flexTeaser.teaser.lead || undefined,
+                    pageID: flexTeaser.teaser.page.id
+                  },
+                  alignment: {
+                    i: flexTeaser?.alignment.i ?? nanoid(),
+                    x: flexTeaser?.alignment.x ?? 1,
+                    y: flexTeaser?.alignment.y ?? 1,
+                    w: flexTeaser?.alignment.w ?? 1,
+                    h: flexTeaser?.alignment.h ?? 1,
+                    static: flexTeaser?.alignment.static ?? false
```
Comment (roosnic1) [inline packages/editor/src/client/blocks/types.ts:484]: i think you don't need the `?` and the `??` here. TypeScript should know about those values.

**Review by roosnic1 — CHANGES_REQUESTED** [review]
Looks really nice and i have some feedback.

File: packages/api/src/graphql/blocks.ts [hunk 24]
```diff
@@ -702,6 +768,21 @@ export const GraphQLTeaserGridBlockInput = new GraphQLInputObjectType({
...
+  fields: {
+    flexTeasers: {type: GraphQLNonNull(GraphQLList(GraphQLFlexTeaserInput))}
```
[the remark this answers is not in the document]
Comment (penk85) [reply packages/api/src/graphql/blocks.ts:812]: @roosnic1 can't it be an empty list?

[the remark this answers is not in the document]
Comment (roosnic1) [reply packages/api/src/graphql/blocks.ts:812]: if you write `GraphQLNonNull(GraphQLList(GraphQLNonNull(GraphQLFlexTeaserInput)))` you're making sure that the list doesn't have any `null`s in it. It still can be an empty list `[]`. The difference is that in the frontend you don't have to check if the items when you iterate over them because you made sure all items of the list are of type `GraphQLFlexTeaserInput` and nothing else.

**Review by camilleory — APPROVED** [review]

**Review by roosnic1 — APPROVED** [review]
🎉 awesome

File: examples/website/src/shared/types.ts (hunk 7 above)
Comment (bishoymelek) [reply examples/website/src/shared/types.ts:349]: @penk85 Will we be using it as blocks? It's a simple point but confusing a little, what do you think?

**Review by bishoymelek — COMMENTED** [review]
🎉 🎉  Great Penina
very small two questions

**Review by bishoymelek — DISMISSED** [review]

**Review by roosnic1 — APPROVED** [review]

**Review by roosnic1 — APPROVED** [review]
👍🏾

**Review by camilleory — APPROVED** [review]

**Review by bishoymelek — APPROVED** [review]
🎉 🎉
